### DISCUSSION TITLE
[Feature] Full Control of External Library or Copy to New Library

### DISCUSSION THREAD

**Discussion by TheDadNerd** [immich-app/immich#5055] [Feature Request]
### The feature

I utilize the icloudpd docker container to pull in my wife and I's iCloud phots on a nightly basis. Main benefit is that it pulls directly from the cloud eliminating our phones from the equation on getting the photos to my server. I have the folder mapped into Immich as an external library so everything shows up which is fantastic. I would love the ability though to delete photos in the external library. Or alternatively I would be fine if Immich watched that external library, like it already does, and copy the items into a new library that I could delete and organize. I know this last option would double up my data as it would leave the external library alone. My use case might be unique but essentially a way to manage external libraries would be a nice opt-in feature for those of us who don't add photos via any Immich UI.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by Daniel-Kalus** [comment]
I would also very much like the option to delete photos from an external library. I don't quite understand why this option is not available. When external library support was added I finally switched to immich, but not being able to delete photos from within the Immich UI kind of defeats the purpose for me to use Immich at all.

**Comment by axis0093** [comment]
I also find this to be a very useful feature. 
Previously, my family and I used Syncthing to synchronize our phone galleries with a NAS,  accumulating a large number of photos. I have set them as external libraries, and Immich does a great job of showcasing these photos. However, when I want to delete some meaningless photos, Immich's external libraries do not allow me to do so.
I have tried using Photoprism, and it supports me in deleting mounted external files. By first setting the photos as archived, I can then delete them within the archived album.

**Comment by Pieeeeeee** [comment]
I also like to be able to add/move/delete pictures and add/move/delete/rename folders.

My current folder structure is divided into years (starting with 2001), every year has subfolders with the name of an activity/event.
What I'd like is being able to move pictures from the upload folder to a folder in the external library. Thus all pictures that are being uploaded via my phone will be evaluated by me (either move them to the external library, preferably by adding new albums on the fly, or delete them if they're crap). So in the end I'll always end up with an empty upload folder.

**Reply by treardon17** [reply]
This is my use case as well. I use Synology photos in this way, where all of my uploads go into a review folder, then I organize/cull the photos until my review folder is empty.

**Reply by aisbergde** [reply]
external libraries are fine with Synology. If you use Synology Drive for photo and video upload, then you can use it as second external library or in the same external library. This works fine, if you don't want to use immich for upload. You can easy move between folders and immich accept this.

I switched from using Synology drive for uploading assets to using immich for uploading assets. I switched, because some edit functionality is planned also in immich and I want to use it. And now I have these issues with moving content between default and external libraries. I have a workaround. But I still hope, that offline asset removing will be implemented also in the default library.

**Reply by treardon17** [reply]
Ideally I would like to move off of Synology photos/drive and use immich for my full use case, mainly because I have a server that has way more resources than my Synology does, but also because I prefer open source solutions. The current implementation of external libraries does not lend itself well to replacing synology photos because it requires an external application (like synology drive, synology photos, nextcloud, etc) to manage the CRUD operations on the content of the external library.

To be honest, if "internal" libraries allowed you to define your own folder structure, that would be perfectly adequate for my use case as well. All I want is a place to put my photos, manage them in a single application, and have control over the naming/structure.

**Reply by Pieeeeeee** [reply]
That is exactly my use case. I have a Synology NAS too, but since it's too old (2013), it became too slow and I don't think it updates anymore. So moving everything to Docker on a much more powerful server, and degrading synology to pure storage.
The problem with Immich indeed is it doesn't have custom folder structure. So for the moment relying on an external library to keep the folder structure.

**Reply by treardon17** [reply]
^^ Exactly. This might go without saying, but I also want to clarify that organization via albums is not sufficient for my use case, because having a preserved folder structure is more conducive to migrating to/from other services (ie. if another cool photo tool comes out, I'm not relying on the immich database), and allows better separation of raw photos and edited photos.

**Comment by aisbergde** [comment]
I also want to combine internal and external libraries in a better way.

Delete support for external libraries would be fine. By moving to a trash folder or by marking assets as deleted in a way, which can be detected from outside. Maybe this can be solved in the future when immich implements a way to export or store metadata in General in a way, other programs can used. Then this could be just a tag (other tags could be: archived, detected faces and so on).

The current main issue with the internal library is the one-way approach: if it is in immich it needs to be maintained by immich and immich don't recognize or accept external changes. I need to be able to delete assets from outside, and I need to be able to modify some metadata from outside (rotation, geo data). After mixing the usage of internal and external libraries I have issues: What I deleted in the internal library from outside is still visible as duplicates in immich. And I am not sure, if external metadata changes will be detected and accepted.

Maybe currently the best way would be to avoid the internal libraries at all. But then immich can't be used for uploads. But I would like to use it for uploading.

**Comment by gitdoc67** [comment]
Having used a immich with external libraries now for a few months, I really wish I had full manageability of these photos within the immich app. It would be great to have a feature in immich to either enabled manageability of these external libraries directly, or to have a migrate button that would copy the photos from the external library to the internal default library while maintaining all thumbnails and ML data.

**Comment by aisbergde** [comment]
IMHO these could be 2 or 3 different steps:

1. marking assets as to be deleted in the database without deleting them physically. The GUI will show them only in the trash bin.

2. physically delete them in similar way, like assets in the default library (after x days), but only, if the library is not read only and if the library has a flag enabled to allow deletings.

3. create a script, containing delete commands per asset, so that an admin can run this script on their own risk independently from immich. (this is just a database query or better a build in view)

**Reply by gitdoc67** [reply]
I think an option 4 of copying the assets to the default library would be best. Using the external library as a true "import" library. Tracking the hash would prevent duplicates and once the file is copied it would be hidden from the external library or optionally deleted.

**Comment by roldengarm** [comment]
I love Immich, and I'm only using it with External Libraries, as I already have a backup / storage solution in place; using QFile all photos are uploaded to the NAS & daily a script runs to put them into year/month/day folder.

I showed Immich to my wife, and first thing she asks: Can I also delete any duplicates?

**Reply by alextran1502** [reply]
All wife's suggestion is valid. This will eventually come. We are just busy working on other priorities at the moment

**Comment by awdwyn** [comment]
My use case ties into this a bit.

I use syncthing to merge the camera roll folders on both my phone and my tablet and also to put a copy on my home server, and then the same for my wife devices. She does her editing on the tablet (cropping, color corrections, deletions, etc). Any changes made are reflected on the server. After a period of time (to give her time to work), older photos/videos in the camera roll folders are moved to a year/month organized master library folder which is not synced to the phones.

I'm currently looking into frontend apps that would make the entire master library (~200k photos and videos) that's on the server, and the separate camera roll folders accessible to mobile apps or a lan accessible website. She needs to be able to organize (via albums) and share-via-link anything that would be in the older master library and also in the camera rolls. I looked into external libraries for this, which work fine for the master library as it is considered mostly-immutable as a self-imposed rule, but fails for the camera rolls.

The problem is that files in the camera roll libraries would be put into albums, shared, etc... and then later moved to the master external library - which immich would see as a bunch of new images. The albums and shared links would stop working. Is there a way to make this work? Does immich track hashes or anything that could be used to recognize that the *new* photo is actually just the *old* photo and change the path internally without breaking the albums/links?

**Comment by aisbergde** [comment]
BTW, the current development wich moves the external library management from users to administrator opens new possibilities. For example hashing assets would allow to detect identical assets in different libraries. These identical assets could be stored only as one asset. Face tagging could be done per asset and automatically be available in all libraries. Moving assets between libraries could keep assets as the are. instead of n:1 relations (many to one) between assets and libraries n:m (many to many relations) could be used.

**Comment by ThreeAurora** [comment]
Yes, I desperately need the ability to remove (or copy, cut, etc.) photos from an external library, otherwise I don't have any way to do my related work

**Reply by Hordemar** [reply]
Here is a little script until this feature is implemented. I cron it twice a day after icloudpd docker finishes pulling photos:

`#!/bin/bash
LOG_FILE="/mnt/exos/immich/logs/$(date +%Y-%m-%d).log"

docker exec Immich immich login-key http://192.168.0.110:8081/api CwCfM3JxsNttrsDzaRGx6rBPh0r3OU6ThRx0heBCyyE >> "$LOG_FILE" 2>&1 && \

sleep 5 && \

docker exec Immich immich upload --recursive /import2 /photos2 >> "$LOG_FILE" 2>&1`

**Comment by Eksistenze** [comment]
+1 to this.  I have icloudpd grabbing all the photos from all the family's iCloud.  I have these locations as external libraries for the user accounts in Immich.  I would really like an option where Immich copies everything into its own library.  I understand that would double up on storage but it would only be temporary in my use case.  I would then empty the external libraries and have a offsite backup for the Immich library.

LOVE the app and I'm in the process of just getting rid of google and apple so I can host my own media.  I would just like this simplified to make backup easier.

**Comment by ThreeAurora** [comment]
Now the feature I need most has been implemented: allowing editing of external libraries. But I see some other good needs in this discussion group.

**Comment by mmomjian** [comment]
This has been implemented in 1.105.0. If you would like a copy of your external pictures, this should be handle outside Immich - for example you can run an hourly cron job to rsync the pictures from one folder to another, and then mount that folder in Immich.
